daemon/containerd/image_push.go

```diff
@@ -2,15 +2,255 @@ package containerd
+					"missing content: %w\n"+
+						"Note: You're trying to push a manifest list/index which "+
+						"references multiple platform specific manifests, but not all of them are available locally "+
+						"or available to the remote repository.\n"+
+						"Make sure you have all the referenced content and try again.",
+					err))
+			}
+		}
+	}
+	return err
+}
+
+// findMissingMountable will walk the target descriptor recursively and return
+// missing contents with their distribution source which could potentially
+// be cross-repo mounted.
+func (i *ImageService) findMissingMountable(ctx context.Context, store content.Store, jobs *jobs,
+	target ocispec.Descriptor, targetRef reference.Named, limiter *semaphore.Weighted,
+) (map[digest.Digest]distributionSource, error) {
+	mountableBlobs := map[digest.Digest]distributionSource{}
+	sources, err := getDigestSources(ctx, store, target.Digest)
+
+	if err != nil {
+		if !errdefs.IsNotFound(err) {
+			return nil, err
+		}
+		logrus.WithField("target", target).Debug("distribution source label not found")
+		return mountableBlobs, nil
+	}
+
+	handler := func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		_, err := store.Info(ctx, desc.Digest)
+		if err != nil {
+			if !cerrdefs.IsNotFound(err) {
+				return nil, errdefs.System(errors.Wrapf(err, "failed to get metadata of content %s", desc.Digest.String()))
+			}
+
+			for _, source := range sources {
+				if canBeMounted(desc.MediaType, targetRef, i.registryService.IsInsecureRegistry, source) {
+					mountableBlobs[desc.Digest] = source
+					jobs.Add(desc)
+					break
+				}
+			}
+			return nil, nil
+		}
+
+		return containerdimages.Children(ctx, store, desc)
+	}
+
+	err = containerdimages.Dispatch(ctx, containerdimages.HandlerFunc(handler), limiter, target)
+	if err != nil {
+		return nil, err
+	}
+
+	return mountableBlobs, nil
+}
+
+func getDigestSources(ctx context.Context, store content.Manager, digest digest.Digest) ([]distributionSource, error) {
+	info, err := store.Info(ctx, digest)
+	if err != nil {
+		if cerrdefs.IsNotFound(err) {
+			return nil, errdefs.NotFound(err)
+		}
+		return nil, errdefs.System(err)
+	}
+
+	sources := extractDistributionSources(info.Labels)
+	if sources == nil {
+		return nil, errdefs.NotFound(fmt.Errorf("label %q is not attached to %s", labelDistributionSource, digest.String()))
+	}
+
+	return sources, nil
+}
+
+// TODO(vvoland): Remove and use containerd const in containerd 1.7+
+// https://github.com/containerd/containerd/pull/8224
+const labelDistributionSource = "containerd.io/distribution.source."
+
+func extractDistributionSources(labels map[string]string) []distributionSource {
+	var sources []distributionSource
+
+	// Check if this blob has a distributionSource label
+	// if yes, read it as source
+	for k, v := range labels {
+		registry := strings.TrimPrefix(k, labelDistributionSource)
+		if registry != k {
+			ref, err := reference.ParseNamed(registry + "/" + v)
```

Comment:
nit: Looks like `registry` collides with the `"github.com/docker/docker/api/types/registry"` import. perhaps `reg` is an option (perhaps also inline it, as it's only used inside the `if`);

```suggestion
		if reg := strings.TrimPrefix(k, labelDistributionSource); reg != k {
			ref, err := reference.ParseNamed(reg + "/" + v)
```

Reply:
Done in: https://github.com/moby/moby/pull/45243